pkg/dockerregistry/server/digestcache.go

```diff
@@ -52,42 +64,122 @@ func (c digestToRepositoryCache) RepositoriesForDigest(dgst digest.Digest) []str
 	return repos.Copy()
 }
 
+// repositoryBucket contains a list of repositories with eviction timeouts.
 type repositoryBucket struct {
 	mu   sync.Mutex
-	list []string
+	list []bucketEntry
 }
 
 // Has returns true if the bucket contains this repository.
-func (i *repositoryBucket) Has(repo string) bool {
-	i.mu.Lock()
-	defer i.mu.Unlock()
-	for _, s := range i.list {
-		if s == repo {
-			return true
+func (b *repositoryBucket) Has(repo string) bool {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := time.Now()
+
+	return b.getIndexOf(repo, &now) >= 0
+}
+
+// Add one or more repositories to this bucket.
+func (b *repositoryBucket) Add(ttl time.Duration, repos ...string) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	now := time.Now()
+	nowPtr := &now
+	evictOn := now.Add(ttl)
+
+	for _, repo := range repos {
+		index := b.getIndexOf(repo, nowPtr)
+		nowPtr = nil // evict just on the first run
+		arr := b.list
+
+		if index >= 0 {
+			// repository already exists, move it to the end with highest eviction time
+			entry := arr[index]
+			copy(arr[index:], arr[index+1:])
+			if entry.evictOn.Before(evictOn) {
+				entry.evictOn = evictOn
+			}
+			arr[len(arr)-1] = entry
+
+		} else {
+			// repository is a new entry
+			if len(arr) >= bucketSize {
```

Comment:
That's possible, but I don't expect all of the repositories to be searched equally. This structure doesn't really fit the way  layers are spread across repositories. There's an empty layer which will be practically in each repository. There will be layers unique to particular repositories as well. Perhaps a priority queue storing pairs of (`blob`, `repository`) would be more fitting.

I don't find it necessary to rework the data structure completely in this PR. Would the current structure be sufficient with maybe some amendments to the size constraints until some follow-up?